File: ios/LookSeeProto/Models/ClusterLandmarkManifest.swift
import Foundation

/// Identifies one immutable cluster-model release.
///
/// `trainingRunId` is treated as the model version.
struct ClusterReleaseKey: Hashable, Codable {
    let clusterId: Int
    let trainingRunId: String
}

/// The cluster-local metadata generated beside `data.yaml` during dataset packaging.
///
/// Migration behavior:
/// - Schema 1 remains readable for already-cached releases.
/// - Schema 2 requires WGS84 latitude and longitude for every landmark.
struct ClusterLandmarkManifest: Codable, Equatable {
    let schemaVersion: Int
    let coordinateSystem: String?
    let clusterId: Int
    let trainingRunId: String
    let generatedAt: String?
    let classCount: Int
    let landmarks: [String: LandmarkManifestEntry]

    var releaseKey: ClusterReleaseKey {
        ClusterReleaseKey(
            clusterId: clusterId,
            trainingRunId: trainingRunId
        )
    }

    var containsRequiredCoordinates: Bool {
        schemaVersion >= 2 &&
        landmarks.values.allSatisfy { $0.hasCoordinates }
    }

    func landmark(for classIndex: Int) -> LandmarkManifestEntry? {
        landmarks[String(classIndex)]
    }

    func validate() throws {
        guard schemaVersion == 1 || schemaVersion == 2 else {
            throw LandmarkManifestValidationError
                .unsupportedSchemaVersion(schemaVersion)
        }

        if schemaVersion == 2 {
            guard coordinateSystem == "WGS84" else {
                throw LandmarkManifestValidationError
                    .invalidCoordinateSystem(
                        expected: "WGS84",
                        actual: coordinateSystem
                    )
            }
        }

        guard !trainingRunId
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .isEmpty
        else {
            throw LandmarkManifestValidationError.emptyTrainingRunId
        }

        guard classCount >= 0 else {
            throw LandmarkManifestValidationError
                .invalidClassCount(classCount)
        }

        guard landmarks.count == classCount else {
            throw LandmarkManifestValidationError
                .classCountMismatch(
                    declared: classCount,
                    actual: landmarks.count
                )
        }

        for classIndex in 0..<classCount {
            let key = String(classIndex)

            guard let entry = landmarks[key] else {
                throw LandmarkManifestValidationError
                    .missingClassIndex(classIndex)
            }

            guard entry.classIndex == classIndex else {
                throw LandmarkManifestValidationError
                    .classIndexMismatch(
                        dictionaryKey: classIndex,
                        entryValue: entry.classIndex
                    )
            }

            guard !entry.landmarkId
                .trimmingCharacters(in: .whitespacesAndNewlines)
                .isEmpty
            else {
                throw LandmarkManifestValidationError
                    .emptyLandmarkId(classIndex: classIndex)
            }

            guard !entry.datasetClassName
                .trimmingCharacters(in: .whitespacesAndNewlines)
                .isEmpty
            else {
                throw LandmarkManifestValidationError
                    .emptyDatasetClassName(classIndex: classIndex)
            }

            guard !entry.label
                .trimmingCharacters(in: .whitespacesAndNewlines)
                .isEmpty
            else {
                throw LandmarkManifestValidationError
                    .emptyDisplayLabel(classIndex: classIndex)
            }

            try validateCoordinates(
                for: entry,
                classIndex: classIndex
            )
        }
    }

    private func validateCoordinates(
        for entry: LandmarkManifestEntry,
        classIndex: Int
    ) throws {
        let hasLatitude = entry.latitude != nil
        let hasLongitude = entry.longitude != nil

        guard hasLatitude == hasLongitude else {
            throw LandmarkManifestValidationError
                .incompleteCoordinates(classIndex: classIndex)
        }

        if schemaVersion == 2 && !hasLatitude {
            throw LandmarkManifestValidationError
                .missingCoordinates(classIndex: classIndex)
        }

        guard
            let latitude = entry.latitude,
            let longitude = entry.longitude
        else {
            return
        }

        guard latitude.isFinite,
              (-90.0...90.0).contains(latitude)
        else {
            throw LandmarkManifestValidationError
                .invalidLatitude(
                    classIndex: classIndex,
                    latitude: latitude
                )
        }

        guard longitude.isFinite,
              (-180.0...180.0).contains(longitude)
        else {
            throw LandmarkManifestValidationError
                .invalidLongitude(
                    classIndex: classIndex,
                    longitude: longitude
                )
        }
    }
}

/// Display and geographic information for one cluster-local class index.
struct LandmarkManifestEntry: Codable, Equatable, Identifiable {
    let classIndex: Int
    let landmarkId: String
    let datasetClassName: String
    let label: String
    let shortDescription: String

    /// Required by schema 2 and absent from legacy schema-1 manifests.
    let latitude: Double?
    let longitude: Double?

    let positiveImageCount: Int?

    var id: String { landmarkId }

    var hasCoordinates: Bool {
        latitude != nil && longitude != nil
    }
}

enum LandmarkManifestValidationError: LocalizedError, Equatable {
    case unsupportedSchemaVersion(Int)
    case invalidCoordinateSystem(expected: String, actual: String?)
    case emptyTrainingRunId
    case invalidClassCount(Int)
    case classCountMismatch(declared: Int, actual: Int)
    case missingClassIndex(Int)
    case classIndexMismatch(dictionaryKey: Int, entryValue: Int)
    case emptyLandmarkId(classIndex: Int)
    case emptyDatasetClassName(classIndex: Int)
    case emptyDisplayLabel(classIndex: Int)
    case incompleteCoordinates(classIndex: Int)
    case missingCoordinates(classIndex: Int)
    case invalidLatitude(classIndex: Int, latitude: Double)
    case invalidLongitude(classIndex: Int, longitude: Double)

    var errorDescription: String? {
        switch self {
        case .unsupportedSchemaVersion(let version):
            return "Unsupported landmark manifest schema version: \(version)."

        case .invalidCoordinateSystem(let expected, let actual):
            return "The landmark manifest requires coordinate system \(expected), but received \(actual) ?? \"no value\")."

        case .emptyTrainingRunId:
            return "The landmark manifest has an empty trainingRunId."

        case .invalidClassCount(let count):
            return "The landmark manifest has an invalid classCount: \(count)."

        case .classCountMismatch(let declared, let actual):
            return "The landmark manifest declares \(declared) classes but contains \(actual) landmark entries."

        case .missingClassIndex(let index):
            return "The landmark manifest is missing class index \(index)."

        case .classIndexMismatch(let dictionaryKey, let entryValue):
            return "The landmark manifest key \(dictionaryKey) contains an entry whose classIndex is \(entryValue)."

        case .emptyLandmarkId(let classIndex):
            return "The landmark at class index \(classIndex) has an empty landmarkId."

        case .emptyDatasetClassName(let classIndex):
            return "The landmark at class index \(classIndex) has an empty datasetClassName."

        case .emptyDisplayLabel(let classIndex):
            return "The landmark at class index \(classIndex) has an empty display label."

        case .incompleteCoordinates(let classIndex):
            return "The landmark at class index \(classIndex) must include both latitude and longitude when either coordinate is present."

        case .missingCoordinates(let classIndex):
            return "The schema-version-2 landmark at class index \(classIndex) is missing latitude and longitude."

        case .invalidLatitude(let classIndex, let latitude):
            return "The landmark at class index \(classIndex) has invalid latitude \(latitude)."

        case .invalidLongitude(let classIndex, let longitude):
            return "The landmark at class index \(classIndex) has invalid longitude \(longitude)."
        }
    }
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/models/ClusterLandmarkManifest.kt
package looksee.angelll.com.models

/** Identifies one immutable cluster-model release. */
data class ClusterReleaseKey(
    val clusterId: Int,
    val trainingRunId: String,
)

/**
 * Cluster-local landmark metadata generated beside the model's data.yaml.
 *
 * The JSON field names intentionally match the iOS/backend contract. A model
 * release and its manifest are always paired by [releaseKey].
 */
data class ClusterLandmarkManifest(
    val schemaVersion: Int,
    val clusterId: Int,
    val trainingRunId: String,
    val generatedAt: String?,
    val classCount: Int,
    val coordinateSystem: String?,
    val landmarks: Map<String, LandmarkManifestEntry>,
) {
    val releaseKey: ClusterReleaseKey
        get() = ClusterReleaseKey(
            clusterId = clusterId,
            trainingRunId = trainingRunId,
        )

    /** Returns the landmark mapped to the model's zero-based class index. */
    fun landmark(classIndex: Int): LandmarkManifestEntry? =
        landmarks[classIndex.toString()]

    /** Validates the mapping before the app registers or activates it. */
    @Throws(LandmarkManifestValidationException::class)
    fun validate() {
        if (schemaVersion != 1 && schemaVersion != 2) {
            throw LandmarkManifestValidationException.UnsupportedSchemaVersion(schemaVersion)
        }

        if (trainingRunId.isBlank()) {
            throw LandmarkManifestValidationException.EmptyTrainingRunId
        }

        if (classCount < 0) {
            throw LandmarkManifestValidationException.InvalidClassCount(classCount)
        }

        if (landmarks.size != classCount) {
            throw LandmarkManifestValidationException.ClassCountMismatch(
                declared = classCount,
                actual = landmarks.size,
            )
        }

        if (schemaVersion == 2 && coordinateSystem != WGS84) {
            throw LandmarkManifestValidationException.InvalidCoordinateSystem(coordinateSystem)
        }

        repeat(classCount) { classIndex ->
            val entry = landmarks[classIndex.toString()]
                ?: throw LandmarkManifestValidationException.MissingClassIndex(classIndex)

            if (entry.classIndex != classIndex) {
                throw LandmarkManifestValidationException.ClassIndexMismatch(
                    dictionaryKey = classIndex,
                    entryValue = entry.classIndex,
                )
            }

            if (entry.landmarkId.isBlank()) {
                throw LandmarkManifestValidationException.EmptyLandmarkId(classIndex)
            }

            if (entry.datasetClassName.isBlank()) {
                throw LandmarkManifestValidationException.EmptyDatasetClassName(classIndex)
            }

            if (entry.label.isBlank()) {
                throw LandmarkManifestValidationException.EmptyDisplayLabel(classIndex)
            }

            if (!entry.latitude.isFinite() || entry.latitude !in -90.0..90.0) {
                throw LandmarkManifestValidationException.InvalidCoordinate(
                    classIndex = classIndex,
                    field = "latitude",
                    value = entry.latitude,
                )
            }

            if (!entry.longitude.isFinite() || entry.longitude !in -180.0..180.0) {
                throw LandmarkManifestValidationException.InvalidCoordinate(
                    classIndex = classIndex,
                    field = "longitude",
                    value = entry.longitude,
                )
            }
        }
    }

    private companion object {
        const val WGS84 = "WGS84"
    }
}

/** Display information associated with one cluster-local class index. */
data class LandmarkManifestEntry(
    val classIndex: Int,
    val landmarkId: String,
    val datasetClassName: String,
    val label: String,
    val shortDescription: String,
    val latitude: Double,
    val longitude: Double,
    val positiveImageCount: Int?,
) {
    val id: String
        get() = landmarkId
}

sealed class LandmarkManifestValidationException(message: String) :
    IllegalArgumentException(message) {

    class UnsupportedSchemaVersion(val version: Int) :
        LandmarkManifestValidationException(
            "Unsupported landmark manifest schema version: $version.",
        )

    data object EmptyTrainingRunId :
        LandmarkManifestValidationException(
            "The landmark manifest has an empty trainingRunId.",
        )

    class InvalidClassCount(val count: Int) :
        LandmarkManifestValidationException(
            "The landmark manifest has an invalid classCount: $count.",
        )

    class ClassCountMismatch(val declared: Int, val actual: Int) :
        LandmarkManifestValidationException(
            "The landmark manifest declares $declared classes but contains " +
                "$actual landmark entries.",
        )

    class MissingClassIndex(val index: Int) :
        LandmarkManifestValidationException(
            "The landmark manifest is missing class index $index.",
        )

    class ClassIndexMismatch(val dictionaryKey: Int, val entryValue: Int) :
        LandmarkManifestValidationException(
            "The landmark manifest key $dictionaryKey contains an entry whose " +
                "classIndex is $entryValue.",
        )

    class EmptyLandmarkId(val classIndex: Int) :
        LandmarkManifestValidationException(
            "The landmark at class index $classIndex has an empty landmarkId.",
        )

    class EmptyDatasetClassName(val classIndex: Int) :
        LandmarkManifestValidationException(
            "The landmark at class index $classIndex has an empty datasetClassName.",
        )

    class EmptyDisplayLabel(val classIndex: Int) :
        LandmarkManifestValidationException(
            "The landmark at class index $classIndex has an empty display label.",
        )

    class InvalidCoordinateSystem(val coordinateSystem: String?) :
        LandmarkManifestValidationException(
            "Schema 2 requires coordinateSystem WGS84; found " +
                "${coordinateSystem ?: "null"}.",
        )

    class InvalidCoordinate(
        val classIndex: Int,
        val field: String,
        val value: Double,
    ) : LandmarkManifestValidationException(
        "The landmark at class index $classIndex has an invalid $field: $value.",
    )
}
